ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -563,17 +567,109 @@ instance Serialise CC.Common.KeyHash where
   encode = toCBOR
   decode = fromCBOR
 
-encodeByronHeader
-  :: Crypto.ProtocolMagicId
-  -> CC.Slot.EpochSlots
-  -> Header (ByronBlockOrEBB cfg)
-  -> Encoding
-encodeByronHeader pm epochSlots (ByronHeaderOrEBB h) = toCBORAHeaderOrBoundary pm epochSlots h
+-- Codec for blocks and headers using CBOR-in-CBOR. This style is needed
+-- because the cardano-ledger decoders drop information that must be
+-- retained if we are to communicate with Byron peers.
+-- Re-annotating in the cardano-ledger style, i.e. re-encoding then using those
+-- bytes to annotate, does not work, because information is lost in the
+-- initial decoding.
+-- This codec must be used for network exchange _and_ for the database.
+
+-- | Get the encoded bytes of a block. A legacy Byron node (cardano-sl) would
+-- successfully decode a block from these.
+blockBytes :: ByronBlockOrEBB cfg -> Lazy.ByteString
+blockBytes blk = Lazy.pack [listLengthByte, discriminatorByte] `Lazy.append` Lazy.fromStrict mainBytes
+  where
+  -- Here's how it would look if we could use CBOR encoding.
+  --
+  --  enc = CBOR.encodeListLen 2 <> blockBytes
+  --  blockEnc = case unByronBlockOrEBB blk of
+  --    CC.Block.ABOBBoundary b -> toCBOR (0 :: Word) <> toCBOR
+  --    CC.Block.ABOBBlock b    -> toCBOR (1 :: Word) <> toCBORBlock epochSlots b
+  --
+  -- But since we don't carry an annotation for the _entire_ block, and we
+  -- can't put a bytestring representing a _part_ of the encoding into an
+  -- encoding, we have to work directly with bytes. Here's how it goes:
```

Comment:
We can do exactly that now. We just need to update package `cborg` to git hash `2e031f0f8570643cb6574c4f7c08cd8ebb3d5135` and use `encodePreEncoded :: B.ByteString -> Encoding`

https://github.com/well-typed/cborg/blob/master/cborg/src/Codec/CBOR/Encoding.hs#L356.

Reply:
Nope. When _decoding_, we need the whole bytestring from which it was decoded, in order to get the proper annotation. No revision of cborg that I'm aware of allows for a `Decoder` to give the actual _bytes_ from which a thing was decoded (only the `ByteSpan`).

Reply:
Oops, I didn't look at the context of this. Yes, we could use the new cborg revision. Is it released yet?

Reply:
It is in master, not yet on hackage. The project files currently refer to a master branch git commit.